### PR TITLE
Adds shopping cart dropdown (WIP)

### PR DESCRIPTION
**Pull request by kul3r4**
Solves #369

### CODE REVIEW

File: src/50_Samples_%26_Templates/Product.html [hunk 1]
```diff
@@ -121,10 +137,34 @@
...
+                <form method="post"
+                  action-xhr="<%host%>/samples_templates/product_listing/preview/add_to_cart"
+                  target="_blank"
+                  on="submit-success:success-lightbox;submit-error:error-lightbox">
+                  <input type="submit" class="button button-primary" value="ADD TO CART">
+                </form>
+            </div>
+            <amp-lightbox id="success-lightbox" class="lightbox" layout="nodisplay">
+                <div class="lightbox-content">
+                    <a href="/shopping_cart/" class="button button-primary call-to-action">Go to Cart</a>
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: Can you use amp-mustache to pass the number of items to the shopping cart page?

Comment (kul3r4) [inline src/50_Samples_%26_Templates/Product.html]: Do you know if it's possible to use amp-mustache inside an amp-lightbox? It doesn't seem to work for me.

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: src/50_Samples_%26_Templates/Product.html [hunk 2]
```diff
@@ -121,10 +137,24 @@
     <body>
     <%>product/header.html%>
 
-        <!-- This button adds the item to the shopping cart. Pressing this button should lead to a the shopping cart page on your normal website and the cart should be prefilled with the current item. -->
+        <!-- This dropdown adds the item to the shopping cart. Pressing the `ADD TO CART` button should lead to a lighbox which will allow to choose whether to continue to the shopping cart page or to continue with your shopping. The shopping cart page will be on your normal website and the cart should be prefilled with the current item. -->
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: Is the lightbox part still relevant?

File: backend/product-listing.go [hunk 3]
```diff
@@ -88,6 +91,24 @@ func registerProductListingHandler(sampleName string) {
...
+	t, err := template.ParseFiles(filePath)
+	if err != nil {
+		panic(err)
+	}
+	http.HandleFunc("/"+sampleName, func(w http.ResponseWriter, r *http.Request) {
+		quantity := r.URL.Query().Get("quantity")
+		price := r.URL.Query().Get("price")
+		img := r.URL.Query().Get("img")
+		name := r.URL.Query().Get("name")
+		header := "<div class=\"shopping-cart-line\"><div class=\"shopping-cart-line-element\"></div><div class=\"shopping-cart-line-element\"></div><div class=\"shopping-cart-line-element\">Price</div><div class=\"shopping-cart-line-element\">Quantity</div>"
```
Comment (sebastianbenz) [inline backend/product-listing.go]: Why define the HTML in a string and not in the template file?

File: src/Shopping_Cart.html [hunk 4]
```diff
@@ -1,21 +1,34 @@
...
 <head>
-  <link rel="canonical" href="<%host%>/amp-form-success/">
+  <link rel="canonical" href="<%host%>/shopping_cart/">
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
   <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
-  <script async src="https://cdn.ampproject.org/v0.js"></script>  
+  <script async src="https://cdn.ampproject.org/v0.js"></script> 
+  <style amp-custom>
+    .shopping-cart-line {
```
Comment (sebastianbenz) [inline src/Shopping_Cart.html]: s/shopping-cart-line/shopping-cart

File: src/Shopping_Cart.html [hunk 5]
```diff
@@ -1,21 +1,34 @@
...
+      display: flex;
+      flex-wrap: wrap;
+    }
+    .shopping-cart-line-element {
```
Comment (sebastianbenz) [inline src/Shopping_Cart.html]: s/shopping-cart-line-element/shopping-cart-item

File: src/50_Samples_%26_Templates/Product.html [hunk 6]
```diff
@@ -79,6 +80,21 @@
         align-items: center;
         justify-content: space-between;
     }
+    .lightbox {
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: remove lightbox stuff

File: src/50_Samples_%26_Templates/Product.html [hunk 7]
```diff
@@ -121,10 +137,24 @@
...
 
-        <!-- This button adds the item to the shopping cart. Pressing this button should lead to a the shopping cart page on your normal website and the cart should be prefilled with the current item. -->
+        <!-- This dropdown adds the item to the shopping cart. Pressing the `ADD TO CART` button should lead to a lighbox which will allow to choose whether to continue to the shopping cart page or to continue with your shopping. The shopping cart page will be on your normal website and the cart should be prefilled with the current item. -->
         <div id="product-header">
             <h2>Apple</h2>
-            <a href="/shopping_cart/" class="button button-primary call-to-action">Buy</a>
+            <div class="call-to-action">
+                <form method="get"                 
+                  action="<%host%>/shopping_cart"
+                  target="_blank">
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: open in same

File: src/Shopping_Cart.html (hunk 4 above)
Comment (sebastianbenz) [inline src/Shopping_Cart.html]: the shopping cart is not visible on mobile

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: src/Shopping_Cart.html [hunk 8]
```diff
@@ -1,21 +1,48 @@
...
+  <!-- #### Shopping Cart -->
+  <!-- An implementation of a Shopping Cart -->
+    <div class="shopping-cart">
+    <div class="shopping-cart-item"></div>
+    <div class="shopping-cart-item"></div>
+    <div class="shopping-cart-item">Price</div>
+    <div class="shopping-cart-item">Quantity</div>
+
+    <div class="shopping-cart-item">
+      <amp-img width="60" height="40" layout="fixed" alt="Apple" src="{{.Img}}"></amp-img>
```
Comment (sebastianbenz) [inline src/Shopping_Cart.html]: Why is the image a template variable but not the alt text?

File: src/50_Samples_%26_Templates/Product.html (hunk 2 above)
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: ?

File: src/50_Samples_%26_Templates/Product.html [hunk 9]
```diff
@@ -79,6 +80,21 @@
...
+            background: rgba(0,0,0,.8);
+    }
+    .lightbox-content {
+            display: flex;
+            flex-direction: row;
+            justify-content: center;
+            padding-top: 16px;
+    }
+    select {
+        height: 36px;
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: move this to styles.css

File: backend/product-listing.go [hunk 10]
```diff
@@ -88,6 +99,22 @@ func registerProductListingHandler(sampleName string) {
...
+	t, err := template.ParseFiles(filePath)
+	if err != nil {
+		panic(err)
+	}
+	http.HandleFunc("/"+sampleName, func(w http.ResponseWriter, r *http.Request) {
+		quantity := r.URL.Query().Get("quantity")
+		price := r.URL.Query().Get("price")
+		img := r.URL.Query().Get("img")
+		name := r.URL.Query().Get("name")
+		t.Execute(w, ShoppingCartProduct{html.UnescapeString(img), name, price, quantity})
```
Comment (sebastianbenz) [inline backend/product-listing.go]: I think a more idiomatic way to create the ShoppingCartProduct would be:

```
ShoppingCartProduct{
  Img: html.UnescapeString(img), 
  Name: r.URL.Query().Get("name")
  ...
}

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: backend/product-listing.go [hunk 11]
```diff
@@ -108,11 +108,13 @@ func registerShoppingCartHandler(sampleName string) {
 		panic(err)
 	}
 	http.HandleFunc("/"+sampleName, func(w http.ResponseWriter, r *http.Request) {
-		quantity := r.URL.Query().Get("quantity")
-		price := r.URL.Query().Get("price")
 		img := r.URL.Query().Get("img")
```
Comment (sebastianbenz) [inline backend/product-listing.go]: I'd inline the image as well

File: src/50_Samples_%26_Templates/Product.html [hunk 12]
```diff
@@ -79,21 +79,12 @@
         align-items: center;
         justify-content: space-between;
     }
-    .lightbox {
-            background: rgba(0,0,0,.8);
-    }
     .lightbox-content {
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: is this needed?

File: src/50_Samples_%26_Templates/Product.html [hunk 13]
```diff
@@ -136,13 +127,13 @@
     <body>
     <%>product/header.html%>
 
-        <!-- This dropdown adds the item to the shopping cart. Pressing the `ADD TO CART` button should lead to a lighbox which will allow to choose whether to continue to the shopping cart page or to continue with your shopping. The shopping cart page will be on your normal website and the cart should be prefilled with the current item. -->
+        <!-- This dropdown allows to select how many items to be added to the shopping cart. Pressing the `ADD TO CART` button should lead to the shopping cart page page. -->
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: s/should lead to the/leads to a/

File: backend/product-listing.go [hunk 14]
```diff
@@ -44,6 +47,13 @@ type Product struct {
 	Url         string `json:"url"`
 }
 
+type ShoppingCartProduct struct {
```
Comment (sebastianbenz) [inline backend/product-listing.go]: suggestion: s/ShoppingCartProduct/ShoppingCartItem/

File: backend/product-listing.go [hunk 15]
```diff
@@ -116,8 +145,7 @@ func searchProducts(sampleName string, query string) ProductListingPage {
 }
 
 func handleSearchRequest(w http.ResponseWriter, r *http.Request, sampleName string) {
-	if r.Method != "POST" {
-		http.Error(w, "post only", http.StatusMethodNotAllowed)
+	if !isFormPostRequest(r.Method, w) {
```
Comment (sebastianbenz) [inline backend/product-listing.go]: why are you no longer returning an error?

File: src/50_Samples_%26_Templates/Product.html [hunk 16]
```diff
@@ -78,6 +79,12 @@
         align-items: center;
         justify-content: space-between;
     }
+    .lightbox-content {
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: remove this

File: src/50_Samples_%26_Templates/Product.html [hunk 17]
```diff
@@ -120,10 +127,24 @@
     <body>
     <%>product/header.html%>
 
-        <!-- This button adds the item to the shopping cart. Pressing this button should lead to a the shopping cart page on your normal website and the cart should be prefilled with the current item. -->
+        <!-- This dropdown allows to select how many items to be added to the shopping cart. Pressing the `ADD TO CART` button should lead to the shopping cart page page. -->
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: s/should lead to the/leads to a/

Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product.html]: Can you move the documentation to the form? I'd might make more sense there. Also mention that you're using amp-form.

File: backend/product-listing.go (hunk 15 above)
Comment (kul3r4) [inline backend/product-listing.go]: the method isFormPostRequest is doing that check

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: backend/product-listing.go (hunk 15 above)
Comment (sebastianbenz) [inline backend/product-listing.go]: ok

**Pull request merged**